### ISSUE TITLE
Blazor WASM: consider adding support for one-file / single-file / atomic deployments

### ISSUE THREAD

**Issue by rodolfograve** [dotnet/runtime#91219]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to deploy a Blazor WASM application to a CDN but given the way a CDN's PoPs are updated (incrementally and non-deterministically), the Blazor WASM application is left in an invalid state for several minutes due to a mix of files from the old and the new version being served by the CDN, which causes integrity check failures - thankfully, otherwise we would just get random behaviour or runtime errors.

This was already requested in [#42088](https://github.com/dotnet/aspnetcore/issues/42088) but rejected because it hurts caching.

To be clear, the ticket above was asking specifically for a "single assembly deployment", which is not what I'm requesting. A single assembly doesn't include HTML, CSS, js and other files that might be part of the system. I don't want to request a specific solution, but I would like the team to think about how can this use case be supported.

For a team like us, deploying to production several times a week, sometimes even several times a day, caching is not a real benefit, and is certainly not a benefit for which we are willing to risk a broken system.

Also, based on our observations, the PublishTrimmed and RunAOTCompilation seem to hurt caching as well, and are already supported. We use them both and when a new version is published most assemblies seem to be downloaded again.

An atomic deployment hurts caching but allows this use case which at the moment is not supported.

Users of the framework, i.e. developers, should be able to choose the approach that works for their situation.

Finally, please notice that we are using Microsoft Azure's CDN, which doesn't have the ability to perform an atomic deployment. I'm not sure if other CDNs can do it, but either way, I still think this is a feature that Blazor WASM could and should support.

### Describe the solution you'd like

A way to deploy a Blazor WASM application as a single file so that deployments that are more complex than "replace files on server with new files" can be performed safely, without ever having a mix of old a new files served to the user.

In particular, deployments to CDNs.

**Comment by lambdageek** [comment]
@rodolfograve Just so I'm totally clear on what you're asking: you'd like the associated HTML, JS, CSS etc files also to all be bundled, not just the .NET assemblies?

**Comment by lambdageek** [comment]
cc @maraf @lewing @pavelsavara

**Comment by lambdageek** [comment]
Possibly duplicate of https://github.com/dotnet/runtime/issues/86162

**Comment by maraf** [comment]
@rodolfograve Would it work for you to always deploy to a unique folder name?

**Comment by rodolfograve** [comment]
@lambdageek that would be ideal since in some cases there are changes to those files that must be in synch with the changes to .NET code. However, I would say a solution that covers only the .NET assemblies would solve 98% of our issues.

I don't want to go too deep into suggesting a solution as I haven't spent enough time thinking about it, but I would say having these options available would be the best possible outcome:

1. Pack all the files together: .NET runtime + _application-specific_**
2. Generate 2 packages, so that they can be cached independently.
2.1. One package for all .NET runtime files, which don't change much between releases.
2.2. One package with all _application-specific_* files.

** I know that having too many options is not necessarily a good thing, so these could be either "always all files (.NET, JS, HTML, CSS, etc). or we could have the option to _include only .NET files_ vs _include all the files_.

I agree that this is closely related to #86162, but that ticket is very specific about JS files. I want to make sure whatever solution comes out of this (if any) focuses on solving the problem of "atomic deployment of a Blazor WASM website to a CDN".

@maraf I'm afraid not. That's what we do today, but even when deploying all the files to a the same/unique folder name, once those files are copied as separate files to a CDN, they get updated individually on each PoP, and the website is left if an inconsistent/broken state for however long it takes for all the files to be updated on all the PoP. Add the fact that many of those "wrong" files will get cached and it can take 15-20 mins for every client to start receiving all the new files.

**Comment by maraf** [comment]
> I agree that this is closely related to #86162, but that ticket is very specific about JS files. I want to make sure whatever solution comes out of this (if any) focuses on solving the problem of "atomic deployment of a Blazor WASM website to a CDN".

That one is not only about JS files, but it also includes dlls (now .wasm files) and other .NET files. What we weren't probably thinking about is wrapping also app specific .js, .css and other standard web files.

> I'm afraid not. That's what we do today, but even when deploying all the files to a the same/unique folder name, once those files are copied as separate files to a CDN, they get updated individually on each PoP, and the website is left if an inconsistent/broken state for however long it takes for all the files to be updated on all the PoP. Add the fact that many of those "wrong" files will get cached and it can take 15-20 mins for every client to start receiving all the new files.

I see. Would it possible to do two step deployment? 
1) Deploy new app version to a unique folder and wait until it propages
2) Deploy entrypoint (probably index.html) pointing to a new version

**Comment by rodolfograve** [comment]
> I see. Would it possible to do two step deployment?
>
> Deploy new app version to a unique folder and wait until it propages
> Deploy entrypoint (probably index.html) pointing to a new version

As far as I know, there is no reliable way to do step 1, i.e. determine when all the files have propagated to all the CDN's PoP ( 
 - short of having a client on every region on the planet, and even that would require deep knowledge of where each PoP is and what criteria is used to route a request to it.

It always helps me when I see a diagram for something, so here's my attempt to add a diagram to explain the problem:

![BlazorCDNFeatureRequest](https://github.com/dotnet/runtime/assets/107422/20aed236-1087-416e-b456-e9424c6753d0)

It looks to me that the solution needs to make sure files are updated on each CDN PoP atomically, so either:

1. Someone adds support for a CDN to update its PoP atomically.
2. Blazor WASM adds support for a single-file deployment, which will make step 4 above be equivalent to an atomic deployment (because a single file is being copied).

Given the complicated infrastructure involved in a CDN, plus the several different providers, I would say step 1 is not a practical solution, which leaves us with option 2 as the only viable option.

**Comment by maraf** [comment]
I'm not familiar with Azure CDN much. Thank you for the details!

The https://github.com/dotnet/runtime/issues/86162 is about bundling all runtime js files into a single one, than bundling dotnet.wasm, dlls, icu, etc into a single wasm (binary) file. So the result would consist of user html, user js and css + single dotnet js and single wasm file. 

As a post build action, user js and dotnet js could be bundled and with css injected into html file. 

The problem remains wasm file, because it's a binary file.

**Comment by rodolfograve** [comment]
Hi there. Should I expect any follow up on this feature request? Let me know if there's anything else you need.

**Comment by rodolfograve** [comment]
Thanks, @maraf.

I have a couple of points to make:

 1. This is ultimately a decision for your team to make (of course), but I don't think bundling solves the problem. Thanks to Blazor (see my point below), I haven't used Java Script as a main language in a couple of years  - so, forgive me if I'm speaking nonsense - but I'm not aware of a bundler that puts together all files (not just js) into a single file. Plus, bundling WASM seems to be a fairly new or experimental feature in both [webpack](https://webpack.js.org/configuration/experiments/) and [rollup](https://www.npmjs.com/package/@rollup/plugin-wasm).
 2. One of the most attractive points of Blazor is that .NET developers don't have to worry about the "messy" JavaScript ecosystem. Asking a .NET developer using Blazor for that reason to use tools like npm, rollup or webpack defeats a good chunk of the advantages of a team that uses Blazor mostly for this reason. Just setting up the development environment in order to use those tools is enough to put back many developers.
 3. This is more a question: what I see in the _framework directory are .dll files (and their compressed versions br and gz). Are those WASM files?

If you're certain that what you propose is a good solution, could I please ask that the outcome of this ticket is some detailed documentation on how to achieve this using existing tools? Perhaps even a project template?

Cheers

**Comment by maraf** [comment]
> Plus, bundling WASM seems to be a fairly new or experimental feature in both [webpack](https://webpack.js.org/configuration/experiments/) and [rollup](https://www.npmjs.com/package/@rollup/plugin-wasm).

I don't think actual bundling of wasm file is possible. HTML/js/css files are text based, WASM file binary. Encoding it into HTML would explode in size. The best think I can think of (as mentioned in previous comment) is ending with two files (html + wasm).

> One of the most attractive points of Blazor is that .NET developers don't have to worry about the "messy" JavaScript ecosystem.

That is fair point, but there are other tools (webpack and rollup are maybe the most complicated ones) like [VS WebCompiler](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.WebCompiler). We believe than introducing another bundler for wasm case (which won't specific in anything when we do https://github.com/dotnet/runtime/issues/86162) is needed and we can spend the resources on other topics that are .net/wasm specific. 

> This is more a question: what I see in the _framework directory are .dll files (and their compressed versions br and gz). Are those WASM files?

In previous versions (.NET < 8) those were standard .NET dlls. In some scenarios they could have been blocked by AV/firewalls and in 8.0 we introduced a "webcil" format, which mitigates these problems. So in .NET 8 you should see `.wasm` files instead of `.dll` (until you explicitly turn it off). More details here https://github.com/dotnet/runtime/issues/80807.

> If you're certain that what you propose is a good solution, could I please ask that the outcome of this ticket is some detailed documentation on how to achieve this using existing tools? Perhaps even a project template?

After https://github.com/dotnet/runtime/issues/86162 is completed, we will produce a documentation or sample app that will showcase it.

**Comment by rodolfograve** [comment]
Thank you, @maraf. Sounds like everything's covered. I'm looking forward to .NET 8 and the new documentation, it will be a huge improvement for us.

**Comment by maraf** [comment]
> Thank you, @maraf. Sounds like everything's covered. I'm looking forward to .NET 8 and the new documentation, it will be a huge improvement for us.

Unfortunately, the single file deployment is scheduled for .NET 9. 
.NET 8 is feature complete and there isn't room for such change to squeeze in.

I'm going to close this issue, as proposed changes are tracked in other issues.
